CMakeLists.txt

```diff
@@ -15,6 +15,10 @@ if(POLICY CMP0141)
   cmake_policy(SET CMP0141 NEW)
 endif()
 
+if (${CMAKE_SOURCE_DIR} STREQUAL ${CMAKE_BINARY_DIR})
+  message(FATAL_ERROR "In-source builds are not allowed.")
```

Comment:
Could we maybe add a less-technical example for first-time users, i.e. "dude, you forgot the `-B build` part"?

Reply:
Mind suggesting the exact wording?

Reply:
I genuinely assume that people read the documentation and build notes before building Bitcoin Core from source :)

So, I took the liberty of leaving the error message as it is.

Reply:
That's not my assumption, but I accept it.

What do you think about my alternative suggestion, i.e.
```
set(CMAKE_DISABLE_SOURCE_CHANGES ON)
set(CMAKE_DISABLE_IN_SOURCE_BUILD ON)
```
instead of our source vs binary comparison?

Reply:
There are 2 reasons:
1. Both variables are not [documented](https://cmake.org/cmake/help/latest/manual/cmake-variables.7.html).
2. The snippet you suggested still does not produce a clear error message:
```
$ cmake -S . -B .
CMake Error at /usr/share/cmake-3.28/Modules/CMakeDetermineSystem.cmake:73 (execute_process):
  execute_process attempted to output into a file: into a source directory.
Call Stack (most recent call first):
  CMakeLists.txt:48 (project)


CMake Error: CMake was unable to find a build program corresponding to "Unix Makefiles".  CMAKE_MAKE_PROGRAM is not set.  You probably need to select a different build tool.
-- Configuring incomplete, errors occurred!
```